<div class="locus-colocalization-result-tab">

  <div class="row p-3 justify-content-center" style="margin-right: 0px; margin-left: 0px;">
      <mat-button-toggle-group 
        name="Results display toggle" 
        aria-label="Choose to display either P-value Correlation plot, HyPrColoc data table, or eCAVIAR data table" 
        [(ngModel)]="selectedResultsDisplay" 
        (change)="changeResultsDisplay()">
        <!-- <mat-button-toggle value="pval_correlation">P-value Correlation</mat-button-toggle> -->
        <mat-button-toggle value="hyprcoloc">HyPrColoc</mat-button-toggle>
        <mat-button-toggle value="ecaviar">eCAVIAR</mat-button-toggle>
      </mat-button-toggle-group>
  </div>

  <!-- <div class="locus-colocalization-pval-correlation" [hidden]="selectedResultsDisplay != 'pval_correlation'">
    <div class="locus-colocalization-content mt-3 px-3">
      <p>
        This scatter plot shows all trait correlation coefficients or P-value calculated from 
        Spearman and Pearson using the -log10(GWAS p-value) and -log10(QTLs p-value) from all variants. 
        The “P-value threshold” can be used to filter variants from QTL and GWAS data before performing 
        correlation analysis.  
      </p>
    </div>
    <div class="row p-3 justify-content-center" style="margin-right: 0px; margin-left: 0px;">
      Toggle Spearman/Pearson correlation type
      <div class="col-md-auto pl-3 pr-3">
        Pearson/Spearman:
        <mat-button-toggle-group 
          name="Pearson / Spearman toggle" 
          aria-label="Choose Pearson or Spearman plot" 
          [(ngModel)]="selectedCorrelation" 
          (change)="triggerCorrelationReplot()">
          <mat-button-toggle value="R">R</mat-button-toggle>
          <mat-button-toggle value="P">P</mat-button-toggle>
        </mat-button-toggle-group>
      </div>
  
      P-value threshold input for correlation scatter plot
      <div 
        class="col-md-auto pl-3 pr-3" 
        id="qtls-locus-colocalization-correlation-scatter-threshold"
        *ngIf="correlationScatter">
        <form [formGroup]="correlationScatterThreshold">
          <mat-form-field class="correlation-pval-threshold-input-field">
            <input 
              #correlationPvalThreshold
              matInput 
              placeholder="P-value Threshold" 
              (keyup)="changeCorrelationPvalThreshold($event)" 
              (keydown.enter)="$event.preventDefault()" 
              [(ngModel)]="selectedCorrelationPvalThreshold" 
              formControlName="correlationPvalThreshold" 
              aria-label="P-value threshold input field">
            <button 
              mat-button matSuffix mat-icon-button aria-label="Clear" 
              (click)="clearCorrelationPvalThreshold()" 
              [disabled]="selectedCorrelation == 'R'">
              <mat-icon>close</mat-icon>
            </button>
            <mat-error *ngIf="correlationScatterThreshold.invalid">{{ correlationPvalThresholdErrorMsg() }}</mat-error>
          </mat-form-field>
        </form>
      </div>
      
    </div>
  
    Show centered Plotly locus colocalization scatter plot if graph is generated
    <div 
      class="d-flex justify-content-center" 
      id="qtls-locus-colocalization-correlation-scatter-plot">
      <plotly-plot 
        *ngIf="correlationScatter"
        [data]="correlationScatter.data" 
        [layout]="correlationScatter.layout" 
        [config]="correlationScatter.config">
      </plotly-plot>
    </div>
  </div> -->

  <!-- <br> -->

  <div class="locus-colocalization-hyprcoloc" [hidden]="selectedResultsDisplay != 'hyprcoloc'">
    <!-- <h3 class="ml-1">HyPrColoc</h3> -->

    <div class="locus-colocalization-content mt-3 px-3">
      <p>
        Hypothesis Prioritization in multi-trait Colocalization (<a href="https://github.com/jrs95/hyprcoloc" target="_blank">HyPrColoc</a>) analyses (Foley <i>et al</i>. 2019 bioRxiv 592238). 
        The first table shows the colocalization results including the following information:  
      </p>
      <ul>
        <li>Traits: a cluster of putatively colocalized traits. The last column shows which gene QTLs are colocalized with GWAS.</li>
        <li>Posterior Probability: the posterior probability that these traits are colocalized</li>
        <li>Regional association probability: always > the posterior probability. Please see Foley <i>et al</i> for details</li>
        <li>Candidate SNP: a candidate causal variant explaining the shared association</li>
        <li>Posterior Explained By SNP: the proportion of the posterior probability explained by the Candidate SNP (which represents the HyPrColoc multi-trait fine-mapping probability).</li>
      </ul>
    </div>

    <!-- Hyprcoloc results table -->
    <div class="hyprcoloc-results pt-2 border border-light rounded" style="background: #FAFAFA !important;">
      <!-- table filter input field -->
      <div class="row ml-2 mr-2">
        <div class="col-10 pl-0">
          <mat-form-field>
              <input matInput 
                [disabled]="!showHyprcolocTable"
                (keyup)="applyFilterHyprcoloc($event.target.value)" 
                placeholder="Filter" 
                aria-label="Hyprcoloc results data table filter input field">
            </mat-form-field>
        </div>
        <div class="col-2 pr-0 pt-4 export-hyprcoloc-details-table" style="color: #0000EE; text-decoration: none; cursor: pointer; text-align: right; vertical-align: middle;">
          <button (click)="exportHyprcolocTable()" 
            *ngIf="showHyprcolocTable" 
            class="pl-0 pr-0"
            style="color: blue; background-color: transparent; border-color: transparent;"
            alt="Export HyPrColoc details table" 
            aria-label="Export HyPrColoc details table">
            Export
        </button>
        </div>
      </div>
      <!-- loading spinner -->
      <div 
        class="d-flex justify-content-center" 
        *ngIf="!showHyprcolocTable && !hyprcolocWarningMessage">
        <mat-spinner></mat-spinner>
      </div>
      <!-- show warning message if no Hyprcoloc data is returned -->
      <div 
        id="hyprcoloc-warning"
        class="ml-2 mr-2 alert alert-warning show text-center" 
        role="alert"  
        *ngIf="hyprcolocWarningMessage">
        No Hyprcoloc calculation results. Please check input data files and calculation parameters.
      </div>
      <!-- x-scrollable table container -->
      <div class="ml-2 mr-2 hyprcoloc-details-table-container" *ngIf="showHyprcolocTable">
        <table class="hyprcoloc-details-table" mat-table [dataSource]="dataSourceHyprcoloc" matSort #HyprcolocTableSort="matSort" matSortActive="posterior_prob" matSortDirection="desc" matSortStart="desc">
          <!--- Note that these columns can be defined in any order.
                The actual rendered columns are set as a property on the row definition" -->
  
          <!-- iteration Column -->
          <ng-container matColumnDef="iteration">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="iteration"> Iterations </th>
            <td mat-cell *matCellDef="let hyprcoloc">{{hyprcoloc.iteration}}</td>
          </ng-container>
        
          <!-- traits Column -->
          <ng-container matColumnDef="traits">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="traits"> Traits </th>
            <td mat-cell *matCellDef="let hyprcoloc">{{hyprcoloc.traits}}</td>
          </ng-container>
        
          <!-- posterior_prob Column -->
          <ng-container matColumnDef="posterior_prob">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="posterior_prob"> Posterior Probability </th>
            <td mat-cell *matCellDef="let hyprcoloc">{{hyprcoloc.posterior_prob}}</td>
          </ng-container>
        
          <!-- regional_prob Column -->
          <ng-container matColumnDef="regional_prob">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="regional_prob"> Regional Probability </th>
            <td mat-cell *matCellDef="let hyprcoloc"> {{hyprcoloc.regional_prob}} </td>
          </ng-container>
        
          <!-- candidate_snp Column -->
          <ng-container matColumnDef="candidate_snp">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="candidate_snp"> Candidate SNP </th>
            <td mat-cell *matCellDef="let hyprcoloc"> {{hyprcoloc.candidate_snp}} </td>
          </ng-container>
        
          <!-- posterior_explained_by_snp Column -->
          <ng-container matColumnDef="posterior_explained_by_snp">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="posterior_explained_by_snp"> Posterior Explained By SNP </th>
            <td mat-cell *matCellDef="let hyprcoloc"> {{hyprcoloc.posterior_explained_by_snp}} </td>
          </ng-container>
        
          <!-- dropped_trait Column -->
          <ng-container matColumnDef="dropped_trait">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="dropped_trait"> Dropped Trait </th>
            <td mat-cell *matCellDef="let hyprcoloc"> {{hyprcoloc.dropped_trait}} </td>
          </ng-container>
        
          <!-- gene_id Column -->
          <ng-container matColumnDef="gene_id">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="gene_id"> Gene ID </th>
            <td mat-cell *matCellDef="let hyprcoloc"> {{hyprcoloc.gene_id}} </td>
          </ng-container>
        
          <!-- gene_symbol Column -->
          <ng-container matColumnDef="gene_symbol">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="gene_symbol">Gene Symbol</th>
            <td mat-cell *matCellDef="let hyprcoloc"> {{hyprcoloc.gene_symbol}} </td>
          </ng-container>
        
          <tr mat-header-row *matHeaderRowDef="displayedColumnsHyprcoloc" style="background: #8C97D4 !important;"></tr>
          <tr mat-row *matRowDef="let row; columns: displayedColumnsHyprcoloc;"></tr>
        </table>
  
      </div>
        
      <mat-paginator #HyprcolocPaginator [pageSizeOptions]="[10, 20, 50, 100]" showFirstLastButtons style="background: #FAFAFA !important;"></mat-paginator>
  
    </div>

    <br>

    <!-- <h3 class="ml-1 mt-5">HyPrColoc SNP Scores</h3> -->

    <div class="locus-colocalization-content mt-3 px-3">
      <p>
        The second table outputs the SNP score of all variants for each pair of colocalized traits. For detail information, please check the manual of <a href="https://github.com/jrs95/hyprcoloc" target="_blank">HyPrColoc</a>.
      </p>
    </div>

    <!-- Hyprcoloc snpscore results table -->
    <div class="hyprcoloc-snpscore-results pt-2 border border-light rounded" style="background: #FAFAFA !important;">
      <!-- table filter input field -->
      <div class="row ml-2 mr-2">
        <div class="col-10 pl-0">
          <mat-form-field>
            <input matInput 
              [disabled]="!showHyprcolocSnpscoreTable"
              (keyup)="applyFilterHyprcolocSnpscore($event.target.value)" 
              placeholder="Filter" 
              aria-label="Hyprcoloc SNP scores data table filter input field">
          </mat-form-field>
        </div>
        <div class="col-2 pr-0 pt-4 export-hyprcoloc-snpscore-details-table" style="color: #0000EE; text-decoration: none; cursor: pointer; text-align: right; vertical-align: middle;">
          <button (click)="exportHyprcolocSnpscoreTable()" 
            *ngIf="showHyprcolocSnpscoreTable" 
            class="pl-0 pr-0"
            style="color: blue; background-color: transparent; border-color: transparent;" 
            alt="Export HyPrColoc SNP score table" 
            aria-label="Export HyPrColoc SNP score table">
            Export
        </button>
        </div>
      </div>
      <!-- loading spinner -->
      <div 
        class="d-flex justify-content-center" 
        *ngIf="!showHyprcolocSnpscoreTable && !hyprcolocSnpscoreWarningMessage">
        <mat-spinner></mat-spinner>
      </div>
      <!-- show warning message if no Hyprcoloc snpscore data is returned -->
      <div 
        id="hyprcoloc-snpscore-warning"
        class="ml-2 mr-2 alert alert-warning show text-center" 
        role="alert"  
        *ngIf="hyprcolocSnpscoreWarningMessage">
        No Hyprcoloc snpscore calculation results. Please check input data files and calculation parameters.
      </div>
      <!-- x-scrollable table container -->
      <div class="ml-2 mr-2 hyprcoloc-snpscore-details-table-container" *ngIf="showHyprcolocSnpscoreTable">
        <table class="hyprcoloc-snpscore-details-table" mat-table [dataSource]="dataSourceHyprcolocSnpscore" matSort #HyprcolocSnpscoreTableSort="matSort" matSortActive="snpscore" matSortDirection="desc" matSortStart="desc">
          <!--- Note that these columns can be defined in any order.
                The actual rendered columns are set as a property on the row definition" -->

          <!-- rsnum Column -->
          <ng-container matColumnDef="rsnum">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="rsnum"> RS Number </th>
            <td mat-cell *matCellDef="let hyprcolocSnpscore">{{hyprcolocSnpscore.rsnum}}</td>
          </ng-container>
        
          <!-- snpscore Column -->
          <ng-container matColumnDef="snpscore">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="snpscore"> SNP Score </th>
            <td mat-cell *matCellDef="let hyprcolocSnpscore">{{hyprcolocSnpscore.snpscore}}</td>
          </ng-container>
        
          <!-- gene_id Column -->
          <ng-container matColumnDef="gene_id">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="gene_id"> Gene ID </th>
            <td mat-cell *matCellDef="let hyprcolocSnpscore">{{hyprcolocSnpscore.gene_id}}</td>
          </ng-container>
        
          <!-- gene_symbol Column -->
          <ng-container matColumnDef="gene_symbol">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="gene_symbol"> Gene Symbol </th>
            <td mat-cell *matCellDef="let hyprcolocSnpscore"> {{hyprcolocSnpscore.gene_symbol}} </td>
          </ng-container>
        
          <tr mat-header-row *matHeaderRowDef="displayedColumnsHyprcolocSnpscore" style="background: #8C97D4 !important;"></tr>
          <tr mat-row *matRowDef="let row; columns: displayedColumnsHyprcolocSnpscore;"></tr>
        </table>

      </div>
        
      <mat-paginator #HyprcolocSnpscorePaginator [pageSizeOptions]="[10, 20, 50, 100]" showFirstLastButtons style="background: #FAFAFA !important;"></mat-paginator>

    </div>
  </div>

  <!-- <br> -->

  <!-- <br> -->

  <div class="locus-colocalization-ecaviar" [hidden]="selectedResultsDisplay != 'ecaviar'">
    <!-- <h3 class="ml-1">eCAVIAR</h3> -->

    <div class="locus-colocalization-content mt-3 px-3">
      <p>
        eCAVIAR is a novel probabilistic model for integrating GWAS and eQTL data that extends the CAVIAR framework to 
        explicitly estimate the posterior probability of the same variant being causal in both GWAS and eQTL studies, 
        while accounting for allelic heterogeneity and LD. This approach can quantify the strength between a causal variant 
        and its associated signals in both studies, and it can be used to colocalize variants that pass the genome-wide significance 
        threshold in GWAS. For detail information. Please check the manuscript of <a href="http://zarlab.cs.ucla.edu/tag/ecaviar/" target="_blank">eCAVIAR</a>.  
      </p>
      <p>
        vQTL perform the eCAVIAR analysis for each gene in QTL data together with GWAS data. Two results will be reported based on the 
        number of SNPs tested: +/- 100 kb (CLPP and Prob_in_pCausalSet) or 50 (CLPP2 and Prob_in_pCausalSet2) SNPs around the GWAS lead SNPs. 
        vQTL combines GWAS, QTL, and eCAVIAR results into one table as shown below. If no QTLs are found for GWAS lead SNP (Lead SNP included=”N”), 
        vQTL will use the nearest variant as a locational proxy of “GWAS lead SNP” for the eCAVIAR analysis. 
      </p>
    </div>

    <!-- eCAVIAR results table -->
    <div class="ecaviar-results pt-2 border border-light rounded" style="background: #FAFAFA !important;">
      <!-- table filter input field -->
      <div class="row ml-2 mr-2">
        <div class="col-10 pl-0">
          <mat-form-field>
            <input matInput 
              [disabled]="!showECAVIARTable"
              (keyup)="applyFilterECAVIAR($event.target.value)" 
              placeholder="Filter" 
              aria-label="eCAVIAR results data table filter input field">
          </mat-form-field>
        </div>
        <div class="col-2 pr-0 pt-4 ecaviar-details-table" style="color: #0000EE; text-decoration: none; cursor: pointer; text-align: right; vertical-align: middle;">
          <button (click)="exportECAVIARTable()" 
            *ngIf="showECAVIARTable" 
            class="pl-0 pr-0"
            style="color: blue; background-color: transparent; border-color: transparent;" 
            alt="Export eCAVIAR details table" 
            aria-label="Export eCAVIAR details table">
            Export
        </button>
        </div>
      </div>
      <!-- loading spinner -->
      <div 
        class="d-flex justify-content-center" 
        *ngIf="!showECAVIARTable && !ecaviarWarningMessage">
        <mat-spinner></mat-spinner>
      </div>
      <!-- show warning message if no eCAVIAR data is returned -->
      <div 
        id="ecaviar-warning"
        class="ml-2 mr-2 alert alert-warning show text-center" 
        role="alert"  
        *ngIf="ecaviarWarningMessage">
        No eCAVIAR calculation results. Please check input data files and calculation parameters.
      </div>
      <!-- x-scrollable table container -->
      <div class="ml-2 mr-2 ecaviar-details-table-container" *ngIf="showECAVIARTable">
        <table class="ecaviar-details-table" mat-table [dataSource]="dataSourceECAVIAR"  matSort #ECAVIARTableSort="matSort" matSortActive="CLPP" matSortDirection="desc" matSortStart="desc">
          <!--- Note that these columns can be defined in any order.
                The actual rendered columns are set as a property on the row definition" -->
  
          <!-- gene_id Column -->
          <ng-container matColumnDef="gene_id">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="gene_id"> Gene ID </th>
            <td mat-cell *matCellDef="let ecaviar">{{ecaviar.gene_id}}</td>
          </ng-container>
        
          <!-- gene_symbol Column -->
          <ng-container matColumnDef="gene_symbol">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="gene_symbol"> Gene Symbol </th>
            <td mat-cell *matCellDef="let ecaviar">{{ecaviar.gene_symbol}}</td>
          </ng-container>
        
          <!-- variant_id Column -->
          <!-- <ng-container matColumnDef="variant_id">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="variant_id"> Variant ID </th>
            <td mat-cell *matCellDef="let ecaviar">{{ecaviar.variant_id}}</td>
          </ng-container> -->
        
          <!-- rsnum Column -->
          <ng-container matColumnDef="rsnum">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="rsnum"> RS Number </th>
            <td mat-cell *matCellDef="let ecaviar">{{ecaviar.rsnum}}</td>
          </ng-container>
        
          <!-- chr Column -->
          <ng-container matColumnDef="chr">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="chr"> Chromosome </th>
            <td mat-cell *matCellDef="let ecaviar"> {{ecaviar.chr}} </td>
          </ng-container>
        
          <!-- pos Column -->
          <ng-container matColumnDef="pos">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="pos"> Position </th>
            <td mat-cell *matCellDef="let ecaviar"> {{ecaviar.pos}} </td>
          </ng-container>
        
          <!-- ref Column -->
          <ng-container matColumnDef="ref">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="ref"> Ref </th>
            <td mat-cell *matCellDef="let ecaviar"> {{ecaviar.ref}} </td>
          </ng-container>
        
          <!-- alt Column -->
          <ng-container matColumnDef="alt">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="alt"> Alt </th>
            <td mat-cell *matCellDef="let ecaviar"> {{ecaviar.alt}} </td>
          </ng-container>
        
          <!-- tss_distance Column -->
          <ng-container matColumnDef="tss_distance">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="tss_distance"> TSS Distance </th>
            <td mat-cell *matCellDef="let ecaviar"> {{ecaviar.tss_distance}} </td>
          </ng-container>
        
          <!-- pval_nominal Column -->
          <ng-container matColumnDef="pval_nominal">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="pval_nominal">Nomimal P-value</th>
            <td mat-cell *matCellDef="let ecaviar"> {{ecaviar.pval_nominal}} </td>
          </ng-container>
        
          <!-- slope Column -->
          <ng-container matColumnDef="slope">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="slope"> Slope </th>
            <td mat-cell *matCellDef="let ecaviar"> {{ecaviar.slope}} </td>
          </ng-container>
        
          <!-- slope_se Column -->
          <ng-container matColumnDef="slope_se">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="slope_se"> Slope SE </th>
            <td mat-cell *matCellDef="let ecaviar"> {{ecaviar.slope_se}} </td>
          </ng-container>
          
          <!-- gwas_pvalue Column -->
          <ng-container matColumnDef="gwas_pvalue">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="gwas_pvalue"> GWAS P-value </th>
            <td mat-cell *matCellDef="let ecaviar"> {{ecaviar.gwas_pvalue}} </td>
          </ng-container>
      
          <!-- gwas_z Column -->
          <ng-container matColumnDef="gwas_z">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="gwas_z"> GWAS Z </th>
            <td mat-cell *matCellDef="let ecaviar"> {{ecaviar.gwas_z}} </td>
          </ng-container>
      
          <!-- Leadsnp Column -->
          <ng-container matColumnDef="Leadsnp">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="Leadsnp"> Lead SNP </th>
            <td mat-cell *matCellDef="let ecaviar"> {{ecaviar.Leadsnp}} </td>
          </ng-container>
          
          <!-- Prob_in_pCausalSet Column -->
          <ng-container matColumnDef="Prob_in_pCausalSet">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="Prob_in_pCausalSet"> Prob_in_pCausalSet </th>
            <td mat-cell *matCellDef="let ecaviar"> {{ecaviar.Prob_in_pCausalSet}} </td>
          </ng-container>
      
          <!-- CLPP Column -->
          <ng-container matColumnDef="CLPP">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="CLPP"> CLPP </th>
            <td mat-cell *matCellDef="let ecaviar"> {{ecaviar.CLPP}} </td>
          </ng-container>
      
          <!-- Prob_in_pCausalSet2 Column -->
          <ng-container matColumnDef="Prob_in_pCausalSet2">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="Prob_in_pCausalSet2"> Prob_in_pCausalSet2 </th>
            <td mat-cell *matCellDef="let ecaviar"> {{ecaviar.Prob_in_pCausalSet2}} </td>
          </ng-container>
      
          <!-- CLPP2 Column -->
          <ng-container matColumnDef="CLPP2">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="CLPP2"> CLPP2 </th>
            <td mat-cell *matCellDef="let ecaviar"> {{ecaviar.CLPP2}} </td>
          </ng-container>
      
          <!-- leadsnp_included Column -->
          <ng-container matColumnDef="leadsnp_included">
            <th mat-header-cell *matHeaderCellDef mat-sort-header="leadsnp_included"> Lead SNP Included </th>
            <td mat-cell *matCellDef="let ecaviar"> {{ecaviar.leadsnp_included}} </td>
          </ng-container>
        
          <tr mat-header-row *matHeaderRowDef="displayedColumnsECAVIAR" style="background: #8C97D4 !important;"></tr>
          <tr mat-row *matRowDef="let row; columns: displayedColumnsECAVIAR;"></tr>
        </table>
  
      </div>
        
      <mat-paginator #ECAVIARPaginator [pageSizeOptions]="[10, 20, 50, 100]" showFirstLastButtons style="background: #FAFAFA !important;"></mat-paginator>
  
    </div>
  </div>

  <br>

</div>