<div class="locus-alignment-result-tab">

  <!-- Popover -->
  <div class="popover">
    <button type="button" class="close" aria-label="Close" (click)="closePopover()">
      <span aria-hidden="true">&times;</span>
    </button>
    <div class="popover-content" *ngIf="popoverData">
      <b>chr{{ popoverData.chr }}:{{ popoverData.pos }}</b>
      <br>
      <span *ngIf="popoverData.rsnum">
        <b>{{ popoverData.rsnum }}</b>
        <br>
      </span>
      Ref/Alt: <b>{{ popoverData.ref }}</b>/<b>{{ popoverData.alt }}</b>
      <br>
      <i>P</i>-value: <b>{{ popoverData.pvalue }}</b>
      <br>
      Slope: <b>{{ popoverData.slope }}</b>
      <br>
      R<sup>2</sup>: <span *ngIf="popoverData.R2"><b>{{ popoverData.R2 }}</b></span><span *ngIf="!popoverData.R2"><b>NA</b></span>
      <hr>
      <a (click)="makeLDRef()">Make LD Reference</a>
      <br>
      <a (click)="linkLDpop()">LDpop</a>
      <br>
      <a (click)="linkGWAS()">Potential GWAS</a>
      <br>
      <a (click)="linkGnomADBrowser()">gnomAD browser</a>
      <br>
      <div id="qtls-enable-locus-alignment-boxplots" *ngIf="!disableLocusQuantification">
        <a (click)="showBoxplot(popoverData)">Show Boxplots</a>
      </div>
      <div id="qtls-disable-locus-alignment-boxplots" class="disabled-overlay" *ngIf="disableLocusQuantification">
        <a>Show Boxplots</a>
      </div>
    </div>
    <div class="popover-content" *ngIf="!popoverData">
      No data
    </div>
  </div>

  <div class="locus-alignment-content mt-3 px-3">
    <p>
      Two <a href="http://locuszoom.org/" target="_blank">LocusZoom</a> plots below show the association <i>P</i>-values for QTLs (top panel), and GWAS (bottom panel). 
      If no GWAS data is uploaded, a single LocusZoom plot will be shown for QTLs only. 
      <br>
      <br>
      On the panel above, 1000 Genomes populations and LD reference SNP can be customized; 
      by default, the LD information (r<sup>2</sup>) will be calculated based on the 1000 Genomes Dataset (phase 3, version 5) 
      but can be replaced by the user uploading their own LD matrix. Use the blue arrowhead on the top left corner to show or hide file uploading menu on the left.
    </p>
    <p>
      The pink diamond in the LocusZoom plot shows the best (most significant) QTL variant for the selected 
      “Reference Gene”.  Red circle indicates the current LD reference variant, which will be simultaneously 
      updated in both GWAS and QTLs LocusZoom plots. 
    </p>
  </div>

  <!-- Show centered Plotly locus alignment manhattan plot(s) if graph is generated -->
  <div 
    class="d-flex justify-content-center" 
    id="qtls-locus-alignment-plot">
    <plotly-plot 
      *ngIf="graph"
      [data]="graph.data" 
      [layout]="graph.layout" 
      [config]="graph.config" 
      [divId]="graph.divId" 
      (click)="clickPoint($event)" 
      (relayout)="closePopover2($event)" 
      (redraw)="closePopover2($event)">
    </plotly-plot>
  </div>

  <!-- <div 
    class="d-flex justify-content-center" 
    id="qtls-locus-alignment-plot">
  </div> -->

  <div class="locus-alignment-content px-3 text-center footnote">
    <p>
      <small>Click on the SNP to show additional information and actions</small>
    </p>
  </div>

  <div class="w-100 mt-1 mb-2 border border-light"></div>

  <div class="locus-alignment-content mt-3 px-3" *ngIf="scatter">
    <p>
      The following scatter plot shows the correlation between -log10(GWAS <i>P</i>-value) and -log10(QTLs <i>P</i>-value). 
      The “<i>P</i>-value threshold” can be used to filter QTL and GWAS variants based on their <i>P</i>-values before visualizing the <i>P</i>-value correlation. 
      For detailed colocalization analyses, please check the “Locus Colocalization” sub-module. 
    </p>
  </div>

  <!-- P-value threshold input for GWAS scatter plot -->
  <div 
    class="d-flex justify-content-center" 
    id="qtls-locus-alignment-scatter-threshold"
    *ngIf="scatter">
    <div class="row">
      <div class="col datafile-title px-0 pt-3 threshold-title">
        P-value Threshold
      </div>
      <div class="col">
        <form [formGroup]="GWASScatterThreshold">
          <mat-form-field class="pval-threshold-input-field" floatLabel="never">
            <input 
              #pvalThreshold
              matInput 
              placeholder="P-value Threshold" 
              (keyup)="changePvalThreshold($event)" 
              (keydown.enter)="$event.preventDefault()" 
              [(ngModel)]="selectedPvalThreshold" 
              formControlName="pvalThreshold" 
              aria-label="P-value threshold input field">
            <button 
              mat-button 
              matSuffix 
              mat-icon-button 
              aria-label="Clear" 
              (click)="clearPvalThreshold()" >
              <mat-icon>close</mat-icon>
            </button>
            <mat-error *ngIf="GWASScatterThreshold.invalid">{{ pvalThresholdErrorMsg() }}</mat-error>
          </mat-form-field>
        </form>
      </div>
    </div>
  </div>

  <!-- Show centered Plotly locus alignment scatter plot if graph is generated -->
  <div 
    class="d-flex justify-content-center" 
    id="qtls-locus-alignment-scatter-plot">
    <plotly-plot 
      *ngIf="scatter"
      [data]="scatter.data" 
      [layout]="scatter.layout" 
      [config]="scatter.config">
    </plotly-plot>
  </div>

  <br>

  <!-- <pre>{{ mainData | json }}</pre> -->

</div>